pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -210,19 +215,97 @@ func (state PipelineRunState) GetNextTasks(candidateTasks map[string]struct{}) [
 	return tasks
 }
 
-// SuccessfulPipelineTaskNames returns a list of the names of all of the PipelineTasks in state
-// which have successfully completed.
-func (state PipelineRunState) SuccessfulPipelineTaskNames() []string {
-	done := []string{}
+// SuccessfulOrSkippedDAGTasks returns a list of the names of all of the PipelineTasks in state
+// which have successfully completed or skipped
+func (state PipelineRunState) SuccessfulOrSkippedDAGTasks(d *dag.Graph) []string {
+	tasks := []string{}
 	for _, t := range state {
-		if t.TaskRun != nil {
-			c := t.TaskRun.Status.GetCondition(apis.ConditionSucceeded)
-			if c.IsTrue() {
-				done = append(done, t.PipelineTask.Name)
+		if _, ok := d.Nodes[t.PipelineTask.Name]; ok {
+			if t.IsSuccessful() || isSkipped(t, state.ToMap(), d) {
+				tasks = append(tasks, t.PipelineTask.Name)
+			}
+		}
+	}
+	return tasks
+}
+
+// isDAGTasksStopped returns true if any of the pipeline task has failed
+// and none of the pipleine task are still running
+func (state PipelineRunState) isDAGTasksStopped(d *dag.Graph) bool {
```

Comment:
I cannot think of any situation where `isDAGTasksStopped` would be true and `isDAGTasksDone` wouldn't.
`isDAGTasksDone` does not imply `isDAGTasksStopped`, but there is no point in the code where we need to check for ``isDAGTasksStopped`  alone either.

Could we remove this?

Reply:
nope, here is an example where `isDAGTasksStopped` returns `true` but `isDAGTasksDone` returns `false`.

```
tasks:
   mytask1 
   mytask6 with success condition 
   mytask8
     runAfter: mytask6
  mytask9
    runAfter: [mytask1, mytask6]
finally:
  mytask2
```,

* mytask1 finishes successfully
* mytask6 starts executing since the condition succeeds but task itself fails

`isDAGTasksStopped`  returns true since `mytask6` has failed and no other tasks are running

`isDAGTasksDone` returns false because there is no way to identify whether `mytask8` would be skipped or not visited. `isSkipped` marks a task skipped if the parent was skipped because of condition failure not because of the task failure.

If you want to I can address this in this PR (I have no idea how without major refactoring 😉 ) or this can be addressed post #2821 The task states are not persisted anywhere and calculated from scratch in `GetPipelineConditionStatus`. Feel free to counter it with any ideas you might have 🙏

Reply:
> nope, here is an example where `isDAGTasksStopped` is `true` but `isDAGTasksDone` is `false`.
> 
> ```
> tasks:
>    mytask1 
>    mytask6 with success condition 
>    mytask8
>      runAfter: mytask6
>   mytask9
>     runAfter: [mytask1, mytask6]
> finally:
>   mytask2
> ```
> 
>     * mytask1 finishes successfully
> 
>     * mytask6 starts executing since the condition succeeds but task itself fails
> 
> 
> `isDAGTasksStopped` returns true since `mytask6` has failed and no other tasks are running
> 
> `isDAGTasksDone` returns false because there is no way to identify whether `mytask8` and `mytask9` would be skipped or not visited. `isSkipped` marks a task skipped if the parent was skipped because of condition failure not because of the task failure.

Yeah, I think that's the fundamental issue :)

> 
> If you want to I can address this in this PR (I have no idea how without major refactoring 😉 ) or this can be addressed post #2821 The task states are not persisted anywhere and calculated from scratch in `GetPipelineConditionStatus`. Feel free to counter it with any ideas you might have 🙏

Reply:
yup 😢 I am working on a PR to clean up `GetPipelineConditionStatus` (merging post this PR) which would hopefully solve this issue

We now consider rest of non-started tasks skipped in case of failure which could be evaluated in `isSkipped()` instead of `GetPipelineConditionStatus`, stay tuned ...